### DISCUSSION TITLE
Error:: Failed to apply cleanly to the shadow database after dropping a column

### DISCUSSION THREAD

**Discussion by ppmaungdevv** [prisma/prisma#21761] [Q&A] [answered]
### Question

I was removing `email column` form User model of the following schema

```
model User {
  id      Int      @id @default(autoincrement())
  email   String   @unique
  name    String?
}
```
to this User model and add a new **Post** model along side
```
model User {
  id      Int      @id @default(autoincrement())
  name    String
  posts   Post[]
}

model Post {
  id         Int      @id @default(autoincrement())
  title      String
  body       String
  user_id    User     @relation(fields: [userId], references: [id])
  userId     Int
  created_at DateTime @default(now())
  updated_at DateTime @default(now())
}
```
The migrate worked fine. Then I updated only the `updated_at` of the **Post** model, do nothing to the User model
```
model Post {
  id         Int      @id @default(autoincrement())
  title      String
  body       String
  user_id    User     @relation(fields: [userId], references: [id])
  userId     Int
  created_at DateTime @default(now())
  updated_at DateTime @updatedAt
}
```
This time when I run the `migrate dev` the following was occurred 

```
Error: P3006

Migration `20231103101523_post` failed to apply cleanly to the shadow database. 
Error code: P3018
Error:
A migration failed to apply. New migrations cannot be applied before the error is recovered from. Read more about how to resolve migration issues in a production database: https://pris.ly/d/migrate-resolve

Migration name: 20231103101523_post

Database error code: 1142

Database error:
INDEX command denied to user 'prisma_user'@'172.22.0.2' for table 'User'

Please check the query number 1 from the migration file.
```
I tried fixing the error by using the methods as shown in [prisma documentation](https://pris.ly/d/migrate-resolve) but it did not work. When I check the query number 1 from the migration file, as show in the error above, I found the Drop query for removing the email index.
```
DROP INDEX `User_email_key` ON `User`;
```
So I delete the query, run the `migrate dev` again then it worked. But I lose all the data.
What I want to know is how are these error related and what is the correct way for updating the column type in a schema without losing the data.
Please, help me out from this situation because I am self learning right now and I been stuck in this problem for too long.

### Information about Prisma Schema, Client Queries and Environment (optional)

```prisma
// Add your schema.prisma
```

```ts
// Add any relevant Prisma Client queries here
```

- OS: 
- Database: 
- Node.js version: 

- Run `prisma -v` to see your Prisma version and paste it

**Comment by ludralph** [comment]
Hi @ppmaungdevv 👋 

Thank you for raising this question.

The error message `INDEX command denied to user 'prisma_user'@'172.22.0.2' for table 'User'` suggests that the database user `prisma_user` does not have the necessary permissions to execute the `DROP INDEX` command on the `User` table. This could be due to the database user not having the necessary permissions to modify the schema of the database. You might need to grant the necessary permissions to the database user.

Secondly, the error message `Migration '20231103101523_post' failed to apply cleanly to the shadow database` indicates that there was an issue applying the migration to the shadow database, which is a temporary database that Prisma Migrate uses to preview the changes that a migration would make to the actual database.

When you manually removed the `DROP INDEX` command from the migration file and ran the migration again, it worked because it no longer tried to execute the command that was causing the error. However, this also resulted in data loss because the `email` column was removed from the `User` model in your Prisma schema, and Prisma Migrate subsequently removed this column from the actual database when you ran the migration.

To avoid data loss when making changes to your Prisma schema, you should follow the steps outlined in the [Prisma Migrate documentation](https://www.prisma.io/docs/guides/migrate/developing-with-prisma-migrate/customizing-migrations) for customizing a migration file. This involves creating a draft migration using the `npx prisma migrate dev  --create-only` command, manually editing the draft migration to ensure that it will make the desired changes to the database without causing data loss, and then applying the migration using the `npx prisma migrate dev` command. In your case, you might need to create a migration that transfers the data from the `email` column to a new column before removing the `email` column.

If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Reply by ppmaungdevv** [reply]
Hi @ludralph 

Thank you for your answers.

I've tried by fixing the permission of the prisma_user as you mentioned but the permission is not the problem. I check my table and the index was deleted successfully. I took the steps in the following ways,

1. remove email column from User model inside `prisma schema` file
2. run `npx prisma migrate dev post` => `success`
3. do some changes in schema file again but didn't change anything inside the User model
4. run `npx prisma migrate dev  --create-only` then I got the following error

```
Error: P3006

Migration `20231103101523_post` failed to apply cleanly to the shadow database. 
Error code: P3018
Error:
A migration failed to apply. New migrations cannot be applied before the error is recovered from. Read more about how to resolve migration issues in a production database: https://pris.ly/d/migrate-resolve

Migration name: 20231103101523_post

Database error code: 1142

Database error:
INDEX command denied to user 'prisma_user'@'172.22.0.2' for table 'User'

Please check the query number 1 from the migration file.
```

From what I understand, if index removing is failed, the error should be showing at `step 2`, right?
Also, I wonder if the steps I've been doing are right or not?

P.S

Permissions of `prisma_user`

```
GRANT CREATE, DROP, REFERENCES, ALTER ON *.* TO `prisma_user`@`%`
GRANT ALL PRIVILEGES ON `mydb`.* TO `prisma_user`@`%`
```

**Reply by philip-roy-jones** [reply]
Thank you for the grant queries!

**Comment by ppmaungdevv** [comment, marked as answer]
I solved the problem. First I delete all my database and user. Then I create a new user for the database with all permission, I am using it for local development so it's not that serious about user permission for now, as answer by the @ludralph. Then I retry all the steps but before I run the `npx prisma migrate dev --create-only`, I run `npx prisma db pull` everytime. I think the two problems are mainly permission for index and the prisma schema was not up to date when I run the migrate command.

**Comment by jpedrosouza** [comment]
The reason for my error was because Prisma creates a shadow database, but since my user only has permission to perform actions on a specific database, when trying to create the INDEX in the shadow database, the error occurred. So to fix it, I simply enabled my user to create INDEX in any schema of my database through the following commands:

`GRANT INDEX ON *.* TO 'your_user'@'%';`
`FLUSH PRIVILEGES;`
